### PR TITLE
[feat] Adding KRISP, additions to OKVQA dataset

### PR DESCRIPTION
**Pull request by KMarino**
Adding KRISP Paper to mmf repo

Additional changes to use image features for OKVQA

### CODE REVIEW

**Review by vedanuj — CHANGES_REQUESTED** [review]
Thanks for adding KRISP to MMF. I have left some comments, this will require some changes before it can land. 

- Please use pre-commit hooks to format the code. You can find details in CONTRIBUTING.md. 
- I think we will have to move around some parts of this PR since this is adding a lot of new dependencies. I would also wait for @apsdehal and @ytsheng reviews and what they think about this.
- Please add more details in your project Readme about the different configurations and how to run them, performance you get on the task for each config.

File: mmf/models/krisp.py
Comment (KMarino) [reply mmf/models/krisp.py:1]: Will do. Going to move all these dependencies to GraphNetworkModule and then try catch that import in krisp

File: mmf/datasets/builders/okvqa/dataset.py
Comment (KMarino) [reply mmf/datasets/builders/okvqa/dataset.py:88]: Will do

File: mmf/models/graphnetwork_module.py
Comment (KMarino) [reply mmf/models/graphnetwork_module.py:23]: Will do

File: mmf/models/visual_bert.py
Comment (KMarino) [reply mmf/models/visual_bert.py:337]: Will do

File: projects/krisp/configs/krisp/okvqa/conceptnet_only.yaml
Comment (KMarino) [reply projects/krisp/configs/krisp/okvqa/conceptnet_only.yaml:1]: How do I do this when I have more than one model (krisp and visual_bert) in my project?

I tried to follow what pretrain_vl_right did (But I do need to get rid of the pre-train subfolder and move those files to the correct places). Please advise

File: projects/krisp/README.md
Comment (KMarino) [reply projects/krisp/README.md]: Fixed

File: mmf/models/krisp.py
Comment (apsdehal) [inline mmf/models/krisp.py:35]: Ideally, you can move your model as well to projects and use user dir to import your model. This would be better approach. @vedanuj what should be better approach do you think is here? It is also not clear in which cases `projects` folder will be registered to python. You may want to `sys.append` manually here so that it is discoverable. Try some other things as well:

- Try pip packaging MMF and installing it to see if you can actually run this.

File: mmf/models/krisp.py
Comment (apsdehal) [inline mmf/models/krisp.py:267]: There is a lot of code duplication going here. Is it super-diverged that you need a custom implementation or can you inherit and get away with it?

**Review by apsdehal — CHANGES_REQUESTED** [review]
Thanks for adding KRISP!

Except for comments for Vedanuj, i have added some additional comments. 
Let me know if something is not clear! We are one step closer to adding KRISP to MMF.

File: mmf/configs/datasets/okvqa/defaults.yaml
Comment (KMarino) [reply mmf/configs/datasets/okvqa/defaults.yaml:19]: Yes, because OKVQA test set overlaps with COCO trainval, so these features are re-trained!

File: mmf/configs/datasets/okvqa/defaults.yaml
Comment (KMarino) [reply mmf/configs/datasets/okvqa/defaults.yaml:26]: Added this to add the features. But yes, we can revert this make this specific to the KRISP files

File: mmf/configs/datasets/okvqa/defaults.yaml
Comment (KMarino) [reply mmf/configs/datasets/okvqa/defaults.yaml:76]: No, I think I copied this from VQA defaults, but it shouldn't be necessary

File: mmf/configs/models/krisp/defaults.yaml
Comment (KMarino) [reply mmf/configs/models/krisp/defaults.yaml:51]: Believe so. It runs and it's the same as in projects/visual_bert/configs/vqa2/defaults.yaml

File: mmf/configs/models/visual_bert/defaults.yaml
Comment (KMarino) [reply mmf/configs/models/visual_bert/defaults.yaml:17]: Yes, I can do that

File: mmf/configs/zoo/datasets.yaml
Comment (KMarino) [reply mmf/configs/zoo/datasets.yaml:368]: The version thing still confuses me. Are you saying make a new dataset heading and don't put this under okvqa?

File: mmf/datasets/builders/okvqa/dataset.py
Comment (KMarino) [reply mmf/datasets/builders/okvqa/dataset.py:27]: There's an extra processing step. And we have essentially two version of OKVQA (v1.0 and v1.1) and only v1.1 is on the okvqa website right now.

I think it's fine for FB to hold its own annotation file for OKVQA, so I don't have a problem with it.

File: mmf/datasets/databases/readers/feature_readers.py
Comment (KMarino) [reply mmf/datasets/databases/readers/feature_readers.py:185]: I thought about this. Made it an if statement because most of the functionality is identical. But might be better to seperate it from an understandability and it makes it less likely this will break other things. I can seperate it if you like.

File: mmf/models/visual_bert.py
Comment (KMarino) [reply mmf/models/visual_bert.py:338]: I can do that. Would that go as seperate file in models/ ?

File: mmf/models/krisp.py
Comment (KMarino) [reply mmf/models/krisp.py:267]: Yeah I know, but there's no way to fix that without rewriting VisualBERT. Need to remove the last part of VisualBERT to make mine work.

File: mmf/configs/datasets/okvqa/defaults.yaml
Comment (apsdehal) [reply mmf/configs/datasets/okvqa/defaults.yaml:19]: I see! Makes sense.

File: mmf/configs/datasets/okvqa/defaults.yaml
Comment (apsdehal) [reply mmf/configs/datasets/okvqa/defaults.yaml:76]: Then please clean this up!

File: mmf/configs/datasets/okvqa/defaults.yaml
Comment (apsdehal) [reply mmf/configs/datasets/okvqa/defaults.yaml:26]: Sure, that would be better.

File: mmf/configs/models/krisp/defaults.yaml
Comment (apsdehal) [reply mmf/configs/models/krisp/defaults.yaml:51]: I think it runs because you override it using the configs in project/visual_bert. Losses don't seem to be on right indentation level. Do confirm once.

File: mmf/configs/zoo/datasets.yaml
Comment (apsdehal) [reply mmf/configs/zoo/datasets.yaml:368]: `defaults` for example is one version. You can add more. Check VizWiz for an example in this file.

File: mmf/datasets/databases/readers/feature_readers.py
Comment (apsdehal) [reply mmf/datasets/databases/readers/feature_readers.py:185]: Yes, you can inherit that feature reader and update this specific part.

File: mmf/models/visual_bert.py
Comment (apsdehal) [reply mmf/models/visual_bert.py:338]: No, it can go directly under your model file `krisp.py` and then you can register it as `visual_bert_krisp` something.

**Review by apsdehal — COMMENTED** [review]
Getting there! Please address the comments and we should be ready to land.

File: mmf/models/krisp.py
Comment (vedanuj) [reply mmf/models/krisp.py:267]: Can you comment here specifically which parts are changed in this version of VisualBERT?

File: mmf/models/krisp.py
Comment (vedanuj) [reply mmf/models/krisp.py:35]: Yes that was my recommendation as well in my previous comments. This model will anyway require additional installation and dependencies so we should move the model under projects folder and use the user directory approach.

Comment (KMarino) [reply mmf/models/krisp.py:35]: I did check and this relative import does work with pip installing. My FB part-time so I'm having to re-set everything up on an external machine, but will give us a sanity check that it works outside facebook.

I'll comment on the remaining comments once I've actually made and pushed those changes. Sorry it's taking me a while, got caught up with other things this week

**Comment by apsdehal** [comment]
@KMarino Any plans on updating this PR? The longer you wait, it will be harder to land with a lot of things changing and moving around.

**Comment by KMarino** [comment]
Yes, got sidetracked. Will try to wrap this up by next week

**Comment by KMarino** [comment]
Yes, got sidetracked. Will try to wrap this up by next week

**Comment by KMarino** [comment]
Made requested changes (I believe I hit them all). Apologies for the delay

File: projects/krisp/README.md
Comment (apsdehal) [inline projects/krisp/README.md:36]: Is this causing issues, this should work well with `env.user_dir` as well.

**Review by apsdehal — APPROVED** [review]
This looks good to go from my side. I will import it now and start the process of landing on Monday after @vedanuj completes a review.

**Comment by apsdehal** [comment]
@KMarino Can you fix failing linters test?

**Comment by KMarino** [comment]
Okay, but it changes a lot of files that have nothing to do with this branch

**Comment by apsdehal** [comment]
You need to rebase on master.

**Comment by KMarino** [comment]
I also seem to have a circular isort black problem with projects/krisp/graphnetwork_module.py and tools/scripts/features/extract_features_vmb.py where if I run black and then isort, the changes cancel each other out.

**Comment by apsdehal** [comment]
Please use `pre-commit run --all` after following contributing guidelines.

**Comment by KMarino** [comment]
Yeah, that is what I'm doing. It does not work because it will endlessly try to reformat those two files but black and isort will undo the changes of the other.

**Comment by KMarino** [comment]
This is my command line

(mmf_release) [kdmarino@compute-1-19]~/code/mmf_fork_pr/mmf:> black projects/krisp/graphnetwork_module.py 
All done! ✨ 🍰 ✨
1 file left unchanged.
(mmf_release) [kdmarino@compute-1-19]~/code/mmf_fork_pr/mmf:> isort projects/krisp/graphnetwork_module.py 
(mmf_release) [kdmarino@compute-1-19]~/code/mmf_fork_pr/mmf:> black projects/krisp/graphnetwork_module.py 
reformatted projects/krisp/graphnetwork_module.py
All done! ✨ 🍰 ✨
1 file reformatted.

The changes undo each other

**Comment by KMarino** [comment]
This is the output of pre-commit run --all

It will output this no matter how many times you run it for the reason I said above

pyupgrade................................................................Passed
isort....................................................................Failed
- hook id: isort
- files were modified by this hook

Fixing tools/scripts/features/extract_features_vmb.py
Fixing projects/krisp/graphnetwork_module.py
Skipped 2 files

black....................................................................Failed
- hook id: black
- files were modified by this hook

reformatted /home/kdmarino/code/mmf_fork_pr/mmf/tools/scripts/features/extract_features_vmb.py
reformatted /home/kdmarino/code/mmf_fork_pr/mmf/projects/krisp/graphnetwork_module.py
All done! ✨ 🍰 ✨
2 files reformatted, 237 files left unchanged.

**Comment by KMarino** [comment]
And now linter tests fail on files that it doesn't on my local pre-commit. For instance black gives me:

would reformat /home/runner/work/mmf/mmf/mmf/datasets/builders/charades/dataset.py
would reformat /home/runner/work/mmf/mmf/mmf/datasets/builders/coco2017/masked_dataset.py
would reformat /home/runner/work/mmf/mmf/mmf/datasets/builders/flickr30k/masked_builder.py
would reformat /home/runner/work/mmf/mmf/mmf/datasets/builders/flickr30k/masked_dataset.py
would reformat /home/runner/work/mmf/mmf/mmf/datasets/builders/coco/detection_dataset.py
would reformat /home/runner/work/mmf/mmf/mmf/datasets/builders/hateful_memes/builder.py
would reformat /home/runner/work/mmf/mmf/mmf/datasets/builders/localized_narratives/masked_builder.py
would reformat /home/runner/work/mmf/mmf/mmf/datasets/builders/mmimdb/builder.py
would reformat /home/runner/work/mmf/mmf/mmf/datasets/builders/localized_narratives/masked_dataset.py
would reformat /home/runner/work/mmf/mmf/mmf/datasets/builders/retrieval/dataset.py
would reformat /home/runner/work/mmf/mmf/mmf/datasets/builders/visual_entailment/builder.py
would reformat /home/runner/work/mmf/mmf/mmf/datasets/builders/visual_genome/detection_builder.py
would reformat /home/runner/work/mmf/mmf/mmf/datasets/builders/visual_genome/masked_builder.py
would reformat /home/runner/work/mmf/mmf/mmf/datasets/mmf_dataset_builder.py
would reformat /home/runner/work/mmf/mmf/mmf/utils/build.py

Aman, I don't even know how I'm supposed to pass these linter tests when I can't even reproduce them locally. Can you please look into this?

**Comment by apsdehal** [comment]
Ok, linter tests are passing now. I will land this in some time.

**Comment by KMarino** [comment]
Thank you @apsdehal!!!

**Comment by ZihaoZheng98** [comment]
What device do you use to train krisp on okvqa?

**Pull request closed without merging**
